// ym-utils/src/simple_router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Route = exports.SimpleRouter = void 0;
var aurelia_framework_1 = require("aurelia-framework");
var SimpleRouter = (function () {
    function SimpleRouter(tq) {
        this.navigation = [];
        this.page_index = 1;
        this.task_queue = tq;
    }
    SimpleRouter.inject = function () { return [aurelia_framework_1.TaskQueue]; };
    ;
    SimpleRouter.prototype.configure = function (navs) {
        this.navigation = navs;
        this.init();
    };
    SimpleRouter.prototype.init = function () {
        var home = this.navigation.find(function (nav) { return nav.route == ''; });
        console.log("init home", home);
        this.navigate(home.name);
    };
    SimpleRouter.prototype.navigate = function (nav_name, param) {
        if (param === void 0) { param = null; }
        console.log("nav...");
        var r = this.navigation.find(function (nav) { return nav.name == nav_name; });
        console.log(r);
        this.current_route = r;
        if (param != null) {
            this.current_route.param = param;
        }
    };
    return SimpleRouter;
}());
exports.SimpleRouter = SimpleRouter;
var Route = (function () {
    function Route() {
    }
    return Route;
}());
exports.Route = Route;
